ledger/acctupdates.go

```diff
@@ -1045,10 +1058,22 @@ func (au *accountUpdates) commitRound(offset uint64, dbRound basics.Round, lookb
 				return
 			}
 		}
+
+		// if everytyhing went well,
+		if err == nil {
+			// take the lock, as we're going to attempt to commit the transaction to database.
+			au.accountsMu.Lock()
+			lockTaken = true
+		}
 		return err
 	})
 
 	if err != nil {
+		// if we failed during the commit, undo the lock.
+		if lockTaken {
+			au.accountsMu.Unlock()
+		}
```

Comment:
can't it be a defer func?

Reply:
it depends where you want to defer this function. In general, no - we want to generate the catchpoint without having the lock taken. Adding a defer function inside this if branch is doable - but not very useful : we're exiting from the function few lines after.